[all hosts] (settings) Update to include recently added Word-specific… (#3977)
OfficeDev/office-js-docs-pr: docs/develop/persisting-add-in-state-and-settings.md

---
title: Persist add-in state and settings
description: Learn how to persist data in Office Web Add-in applications running in the stateless environment of a browser control.
ms.date: 04/25/2023
ms.localizationpriority: medium
---

# Persist add-in state and settings

[!include[information about the common API](../includes/alert-common-api-info.md)]

> Office Add-ins are essentially web applications running in the stateless environment of a browser control. As a result, your add-in may need to persist data to maintain the continuity of certain operations or features across sessions of using your add-in. For example, your add-in may have custom settings or other values that it needs to save and reload the next time it's initialized, such as a user's preferred view or default location. To do that, you can:

Reviewer: ```suggestion
Office Add-ins are essentially web applications running in the stateless environment of a browser control. When in use your add-in may need to persist data to maintain the continuity of certain operations or features across browser sessions. For example, your add-in may have custom settings or other values that it needs to save and reload the next time it's initialized, such as a user's preferred view or default location. To do that, you can:
```

>  The setting with the specified name is created if it doesn't already exist, or its value is updated if it does exist. Use the `Settings.saveAsync` method to persist the new or updated settings to the document.

Reviewer: ```suggestion
The setting with the specified name is created if it doesn't already exist, or its value is updated if it does exist. Use the `Settings.saveAsync` method to persist the new or updated settings to the document.
```

## How to save custom XML to the document

> [!NOTE]
> This section discusses custom XML parts in the context of the Office Common JavaScript API which is supported in Word. The application-specific JavaScript APIs for Excel and for Word also provide access to the custom XML parts. The application-specific APIs and programming patterns are somewhat different from the Common version. For more information, see [Excel.CustomXmlPart](/javascript/api/excel/excel.customxmlpart) and [Word.CustomXmlPart](/javascript/api/word/word.customxmlpart).

> There's an additional storage option when you need to store information that exceeds the size limits of the document settings or which has a structured character. You can persist custom XML markup in a task pane add-in for Word (and for Excel and Word using application-specific API per the note at the beginning of this section). In Word, you can use the [CustomXmlPart](/javascript/api/office/office.customxmlpart) object and its methods (again, see the earlier note about application-specific APIs for Excel and Word). The following code creates a custom XML part and displays its ID and then its content in divs on the page. Note that there must be an `xmlns` attribute in the XML string.

Reviewer: ```suggestion
There's an additional storage option when you need to store information that exceeds the size limits of the document settings or which has a structured character. You can persist custom XML markup in a task pane add-in for Word (and for Excel and Word using application-specific API as mentioned in the previous paragraph). In Word, you can use the [CustomXmlPart](/javascript/api/office/office.customxmlpart) object and its methods. The following code creates a custom XML part and displays its ID and then its content in divs on the page. Note that there must be an `xmlns` attribute in the XML string.
```
I think two mentions of the previous note is unnecessary. If you are concerned that people might skip over the note, consider changing it to a paragraph and then say "as mentioned in the previous paragraph" instead.

}
```

> To retrieve a custom XML part, use the [getByIdAsync](/javascript/api/office/office.customxmlparts#office-office-customxmlparts-getbyidasync-member(1)) method, but the ID is a GUID that is generated when the XML part is created, so you can't know when coding what the ID is. For that reason, it's a good practice when creating an XML part to immediately store the ID of the XML part as a setting and give it a memorable key. The following method shows how to do this. (But see earlier sections of this article for details and best practices when working with custom settings.)

Reviewer: ```suggestion
To retrieve a custom XML part, use the [getByIdAsync](/javascript/api/office/office.customxmlparts#office-office-customxmlparts-getbyidasync-member(1)) method, but the ID is a GUID that is generated when the XML part is created, so you can't know when coding what the ID is. For that reason, it's a good practice when creating an XML part to immediately store the ID of the XML part as a setting and give it a memorable key. The following method shows how to do this.
```
